### DISCUSSION TITLE
Combining Stateful and Partial

### DISCUSSION THREAD

**Discussion by Natureshadow** [winnow-rs/winnow#580] [Q&A] [answered]
I am having a hard time combining the `Stateful` and `Partial` stream types. What I ultimately want to do is stream parsing, but with parser state.

How can I combine `Partial` and `Stateful` correctly?

I tried two approaches:

## Partial wrapping Stateful

```rust
pub type Stream<'i> = Partial<Stateful<&'i str, State<'i>>>;
```

This worked out quite well so far, until I used a `State` type that I cannot (or rather, don't want to) implement `Clone` on:

```plain
error[E0277]: the trait bound `parsers::State<'_>: Clone` is not satisfied
   --> framework/src/sources/graphs/n_triples.rs:61:35
    |
61  |             match triple_parser::<InputError<Stream<'_>>>.parse_peek(input.clone()) {
    |                                   ^^^^^^^^^^^^^^^^^^^^^^ the trait `Clone` is not implemented for `parsers::State<'_>`, which is required by `Partial<Stateful<&str, parsers::State<'_>>>: Clone`
    |
    = note: required for `Stateful<&str, parsers::State<'_>>` to implement `Clone`
    = note: 1 redundant requirement hidden
    = note: required for `Partial<Stateful<&str, parsers::State<'_>>>` to implement `Clone`
note: required by a bound in `InputError`
   --> /home/nik/.cargo/registry/src/index.crates.io-6f17d22bba15001f/winnow-0.6.18/src/error.rs:354:26
    |
354 | pub struct InputError<I: Clone> {
    |                          ^^^^^ required by this bound in `InputError`
```

I assume this is because `Partial`, which I used like shown in the tutorial, wants to copy the input stream to the error.

## Stateful wrapping Partial

```rust
pub type Stream<'i> = Stateful<Partial<&'i str>, State<'i>>;
```

This yields the same `Clone` issue, plus I now cannot use other methods anymore either:

```plain
error[E0308]: mismatched types
  --> framework/src/sources/graphs/n_triples.rs:61:70
   |
61 |             match triple_parser::<InputError<Stream<'_>>>.parse_peek(input.clone()) {
   |                                                           ---------- ^^^^^^^^^^^^^ expected `Stateful<Partial<&str>, State<'_>>`, found `Partial<&str>`
   |                                                           |
   |                                                           arguments to this method are incorrect
   |
   = note: expected struct `Stateful<Partial<&_>, parsers::State<'_>>`
              found struct `Partial<&_>`
```

**Comment by epage** [comment, marked as answer]
`InputError` clones your Input.  Is there a reason you are using that error type, rather than a different one like `ContextError` (the default)?

Also, is there a reason you are using `parse_peek`?  That generally forces you into cloning your input which you likely don't want to be doing.  It mostly exists to help people migrate from `nom` (where the input is cloned all the time).  Can you use `peek()` instead?

**Reply by Natureshadow** [reply]
The answer to both questions is solely that I copied from the docs: https://docs.rs/winnow/0.6.18/winnow/_topic/partial/index.html

Both the error types and the use of `parse_peek` are taken from the provided example.

**Reply by epage** [reply]
I've posted #581 for switching to `ContextError` (got overlooked as we slowly moved further and further away from `InputError`). and #582 for `parse_peek` (looks like it was the only production case in the examples we missed)>

**Reply by praveenperera** [reply]
I'm a little confused about when to use `parse_peek` vs `parse_next`, does the cloning only happen on error?

I am making a streaming parser, and in case of `Incomplete` I want the bytes to be returned unchanged. So I can collect more bytes and try again. In this case `parse_peek` is probably my best bet right?

**Reply by epage** [reply]
We cover parsing partial input at https://docs.rs/winnow/latest/winnow/_topic/partial/index.html

`parse_peek` is mostly there to help people transition from older versions and to make it easier to test parsers.

**Comment by Natureshadow** [comment]
With winnow 0.7 and the migration to `ModalResult`, de facto dictating `ContextError`, the issue has gone away.

**Reply by epage** [reply]
To be clear, we migrated to `ModalResult` being optional and `ContextError` is just as optional as it ever was (a default error type but not required)
